src/data/sym/basic.lean

```diff
@@ -314,6 +314,54 @@ multiset.mem_attach _ _
   :=
 coe_injective $ multiset.attach_cons _ _
 
+@[simps] protected def cast {n m : ℕ} (h : n = m) : sym α n ≃ sym α m :=
+{ to_fun := λ s, ⟨s.val, s.2.trans h⟩, inv_fun := λ s, ⟨s.val, s.2.trans h.symm⟩,
+  left_inv := λ s, subtype.ext rfl, right_inv := λ s, subtype.ext rfl }
+
+@[simp] lemma cast_rfl : sym.cast rfl s = s := subtype.ext rfl
+
+@[simp] lemma cast_cast {n n' n'' : ℕ} (s : sym α n) (h : n = n') (h' : n' = n'') :
+  sym.cast h' (sym.cast h s) = sym.cast (h.trans h') s := rfl
+
+@[simp] lemma coe_cast {n m : ℕ} (h : n = m) (s : sym α n) :
+  (sym.cast h s : multiset α) = s := rfl
+
+def append {n n' : ℕ} (s : sym α n) (s' : sym α n') : sym α (n + n') :=
+⟨s.1 + s'.1, by simp_rw [← s.2, ← s'.2, map_add]⟩
+
+@[simp] lemma append_inj_right {n n' : ℕ} (s : sym α n) (t t' : sym α n') :
+  s.append t = s.append t' ↔ t = t' :=
+subtype.ext_iff.trans $ (add_right_inj _).trans subtype.ext_iff.symm
+
+@[simp] lemma append_inj_left {n n' : ℕ} (s s' : sym α n) (t : sym α n') :
+  s.append t = s'.append t ↔ s = s' :=
+subtype.ext_iff.trans $ (add_left_inj _).trans subtype.ext_iff.symm
+
+lemma append_comm {n n' : ℕ} (s : sym α n) (s' : sym α n') :
+  s.append s' = sym.cast (add_comm _ _) (s'.append s) :=
+by { ext, simp [append, add_comm], }
+
+/-- fill an elment `sym α (n - i)` with `i` copies of `a` to obtain an element of `sym α n` -/
+def fill (a : α) {i : ℕ} (h : i ≤ n) (m : sym α (n - i)) : sym α n :=
+sym.cast (nat.sub_add_cancel h) (m.append (repeat a i))
+
+/-- filter an element of `sym α n` of `a`, if it has `i` copies,
+  it yields an element of `sym α (n - i)` -/
+def filter_ne [decidable_eq α] (a : α) {n : ℕ} (m : sym α n) : Σ i : fin (n + 1), sym α (n - i) :=
+⟨⟨m.1.count a, (multiset.count_le_card _ _).trans_lt $ by rw [m.2, nat.lt_succ_iff]⟩,
+  sym.mk (m.1.filter ((≠) a)) $ eq_tsub_of_add_eq begin
+    conv_rhs { rw [← m.2, multiset.card_eq_countp_add_countp ((=) a), add_comm] },
+    rw multiset.countp_eq_card_filter, refl,
+  end⟩
+
+lemma sigma_ext (m₁ m₂ : Σ i : fin (n + 1), sym α (n - i)) (h : m₁.2.1 = m₂.2.1) : m₁ = m₂ :=
+sigma.subtype_ext (fin.ext $ begin
+  have h₁ := nat.sub_sub_self (nat.lt_succ_iff.1 m₁.1.2),
+  have h₂ := nat.sub_sub_self (nat.lt_succ_iff.1 m₂.1.2),
+  dsimp only [fin.val_eq_coe] at h₁ h₂,
+  rw [← h₁, ← h₂, ← m₁.2.2, ← m₂.2.2, h],
+end) h
```

Comment:
The `h` argument should be in simp normal form, which means that rather than using `.1` (i.e., `subtype.val`) it should use a coercion. We can also golf the proof a bit:
```suggestion
lemma sigma_ext (m₁ m₂ : Σ i : fin (n + 1), sym α (n - i))
  (h : (m₁.2 : multiset α) = m₂.2) : m₁ = m₂ :=
sigma.subtype_ext (fin.ext $ by rw [← nat.sub_sub_self m₁.1.is_le, ← nat.sub_sub_self m₂.1.is_le,
  ← m₁.2.2, ← m₂.2.2, subtype.val_eq_coe, subtype.val_eq_coe, h]) h
```
The main thing in here is using `nat.sub_sub_self m₁.1.is_le` rather than `nat.sub_sub_self (nat.lt_succ_iff.1 m₁.1.2)`.

I'm not sure what to call this lemma -- maybe `sigma_sub_ext`?